contracts/rewards/src/contract/execute.rs

```diff
@@ -45,12 +46,54 @@ where
 
     pub fn record_participation(
         &mut self,
-        _event_id: String,
-        _worker: Addr,
-        _contract: Addr,
-        _block_height: u64,
+        event_id: nonempty::String,
+        worker: Addr,
+        contract: Addr,
+        block_height: u64,
     ) -> Result<(), ContractError> {
-        todo!()
+        let cur_epoch = self.get_current_epoch(block_height)?;
+
+        let event = self
+            .store
+            .load_event(event_id.to_string(), contract.clone())?;
+
+        let mut tally = match event {
+            Some(event) => self
+                .store
+                .load_epoch_tally(contract.clone(), event.epoch_num)?
+                .expect("couldn't find epoch tally for existing event"),
+            None => {
+                self.store.save_event(&Event {
+                    event_id,
+                    contract: contract.clone(),
+                    epoch_num: cur_epoch.epoch_num,
+                })?;
+
+                self.store
+                    .load_epoch_tally(contract.clone(), cur_epoch.epoch_num)?
+                    .map_or(
+                        EpochTally {
+                            event_count: 1,
+                            participation: HashMap::new(),
+                            contract,
+                            epoch: cur_epoch,
+                        },
+                        |tally| EpochTally {
+                            event_count: tally.event_count + 1,
+                            ..tally
+                        },
+                    )
```

Comment:
```suggestion
                self.store
                    .load_epoch_tally(contract.clone(), cur_epoch.epoch_num)?
                    .unwrap_or(EpochTally {
                        contract,
                        event_count: 0,
                        participation: HashMap::new(),
                        epoch: cur_epoch,
                    })
                    .increment_event_count()
```
Extract the increment of `event_count` into a method of `EpochTally`.

Reply:
I'd also prefer a `new` function for the creation of the `EpochTally`. The multi-line create of structure is just ugly.